pulsar-client-shaded/pom.xml

```diff
@@ -91,6 +91,7 @@
 
               <artifactSet>
                 <includes>
+                  <include>org.apache.pulsar:pulsar-client-api</include>
```

Comment:
I think the problem here will be the `pulsar-client` module (which is the shaded module by default) will included the `pulsar-client-api` instead of depending on it. This will make that `pulsar-client` code will not have javadoc or source jars attached, making the IDE experience much worse.

Reply:
@merlimat

I would think that being able to use pulsar client with a recent version of java that has the module system enabled is important. There are ways to get around this though. Here are some alternatives:

1. Accept this PR and a poorer IDE experience. Users can still use the non-shaded original artifacts. Maybe also make a source jar for the shaded jar for the pulsar parts of the code.

2. Fix the split package issue once and for all by moving the classes in the packages that are affected by split-package problem in the pulsar-client-api (or pulsar-common) module to another package. This would be the long-term preferred solution, and the only reason for not doing this would be to maintain backwards-compatibility for external clients.

3. Duplicate the pulsar-client-shaded and pulsar-client-admin-shaded modules and rename the duplicates with a suffix calling them for example pulsar-client-jpms and pulsar-client-admin-jpms or something else, and then publish these on maven central. This is very similar to what is done with the existing pulsar-client-all module which has not been published to maven central so far, and that brings me to the next alternative.

4. Publish pulsar-client-all on maven central along with all the other pulsar client modules. This provides at least one shaded module that can be used properly with recent java clients with jpms enabled.

Reply:
> I would think that being able to use pulsar client with a recent version of java that has the module system enabled is important.

Agree 100% with that 

> Accept this PR and a poorer IDE experience. Users can still use the non-shaded original artifacts. Maybe also make a source jar for the shaded jar for the pulsar parts of the code.

Sure, we had that before 2.3 release, in which we split out the `pulsar-client-api` module explicitly for keeping it unshaded and allow to attach the source/javadoc there.

> Fix the split package issue once and for all by moving the classes in the packages that are affected by split-package problem in the pulsar-client-api (or pulsar-common) module to another package. This would be the long-term preferred solution, and the only reason for not doing this would be to maintain backwards-compatibility for external clients.

Agree that this should be the preferred solution. Just to make sure that I understand the issue: the is that we have classes in the same package name that are present in multiple jars. 

We could fix the issue by either: 
 * Move the classes so that 1 package is in only one jar
 * Do the moving at shading time? 

Do you have a list of the problematic packages and artifacts that create the conflict?

> Duplicate the pulsar-client-shaded and pulsar-client-admin-shaded modules and rename the duplicates with a suffix calling them for example pulsar-client-jpms and pulsar-client-admin-jpms or something else, and then publish these on maven central. This is very similar to what is done with the existing pulsar-client-all module which has not been published to maven central so far, and that brings me to the next alternative.

If we could do that without include `pulsar-client-api` we we'd be good. 

> Publish pulsar-client-all on maven central along with all the other pulsar client modules. This provides at least one shaded module that can be used properly with recent java clients with jpms enabled.

I'm actually surprised the `pulsar-client-all` is not present on Maven. I thought this was published since 2.3 release..

Reply:
> Agree that this should be the preferred solution. Just to make sure that I understand the issue: the is that we have classes in the same package name that are present in multiple jars.
> 
> We could fix the issue by either:
> 
> * Move the classes so that 1 package is in only one jar
> * Do the moving at shading time?
> 
> Do you have a list of the problematic packages and artifacts that create the conflict?
> 

Your understanding of the issue is correct. The artifacts/jars/modules that have a conflict are pulsar-client-api and pulsar-common. Both these modules contain the two packages org.apache.pulsar.common.api and org.apache.pulsar.common.schema.

Moving the classes so that these packages are in only one jar would be preferred.

If you move the classes at shading time (that would be in pulsar-common then since pulsar-client-api is a module you want to keep outside shaded jar), you will still break backwards compatibility with external clients, hence you might just as well move classes from either module in the source tree breaking backwards compatibility but solving the issue permanently and for all artifacts including the non-shaded ones.

Reply:
And also, moving externally visible pulsar classes at shading time would make the pulsar api different and incompatible between the shaded and non-shaded artifacts, and I think that would also be problematic.

Reply:
Got it. I think there should be a simple way out: 

 1. the classes in `org.apache.pulsar.common.api` pkg in `pulsar-common` are not really part of public API. It's just very internal stuff (low-level protocol serialization and so on..). I believe there should be no problem in renaming that package.

 2. Regarding `org.apache.pulsar.common.schema`. It contains POJO definitions that are mostly used in the Admin API. One option would be to move these classes into `pulsar-client-api`, since they're not having any other dependencies.

Reply:
Sounds like an excellent idea! I will perform the changes and push them here asap!

Reply:
Actually, the SchemaInfoUtil class in the org.apache.pulsar.common.schema package from the pulsar-common module depends on classes Commands from package org.apache.pulsar.common.api and PulsarApi from package org.apache.pulsar.common.api.proto.